crates/turborepo-repository/src/package_manager/mod.rs

```diff
@@ -483,7 +484,49 @@ impl PackageManager {
             &globs.validated_exclusions,
             globwalk::WalkType::Files,
         )?;
-        Ok(files.into_iter())
+
+        // we need to remove package.json files that are in subfolders of others so that
+        // we don't yield subpackages. sort, keep track of the parent of last
+        // json we encountered, and only yield it if it's not a subfolder of it
+        //
+        // ideally we would do this during traversal, but walkdir doesn't support
+        // inorder traversal so we can't
+        let mut last_parent = None;
+        Ok(files
+            .into_iter()
+            .sorted_by(|a, b| {
+                // get an iterator of the components of each path, and zip them together
+                let mut segments = a.components().is_last().zip(b.components().is_last());
+
+                // find the first pair of components that are different, and compare them
+                // if one of the segments is the last, then the other is a subfolder of it
+                segments
+                    .find_map(|((a_last, a_cmp), (b_last, b_cmp))| {
+                        if a_last == b_last {
+                            match a_cmp.cmp(&b_cmp) {
+                                std::cmp::Ordering::Equal => None,
+                                other => Some(other),
+                            }
+                        } else if a_last {
+                            Some(std::cmp::Ordering::Less)
+                        } else {
+                            Some(std::cmp::Ordering::Greater)
+                        }
+                    })
+                    .unwrap_or(std::cmp::Ordering::Equal)
```

Comment:
If I'm understanding this, we want to lexicographically order the paths. If so I think [`Iterator::cmp`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.cmp) is what we want. If don't want lexicographic ordering, then we need a comment explaining the ordering.
```suggestion
                a.components().cmp(b.components())
```

Reply:
It's not quite the same. Maybe the comment wasn't clear. The issue is that we want items without subfolders (ie files) to appear before subfolders and their contents. Just doing lexicographical ordering doesn't cut it since you can end up with folder `package/a/b/c/package.json` before `package/package.json`, so we need to check the special case if it is the last segment and always put those first